src/connection.rs

```diff
@@ -0,0 +1,462 @@
+        let stream_id = frame.header().id();
+        if frame.header().flags().contains(ACK) { // pong
+            Ok(None) // TODO
+        } else if self.streams.contains_key(&stream_id) {
+            let mut hdr = Header::ping(frame.header().nonce());
+            hdr.ack();
+            Ok(Some(Frame::new(hdr)))
+        } else {
+            debug!("received ping for unknown stream {}", stream_id);
+            Ok(None)
+        }
+    }
+
+    fn on_go_away(&mut self, frame: &Frame<GoAway>) {
+        debug!("received go_away frame; error code = {}", frame.header().error_code());
+        self.terminate()
+    }
+
+    fn on_reset(&mut self, id: stream::Id) {
+        self.deliver(id, Item::Reset);
+        self.streams.remove(&id);
+    }
+
+    fn on_finish(&mut self, id: stream::Id) {
+        self.deliver(id, Item::Finish);
+    }
+
+    fn next_stream_id(&mut self) -> Result<stream::Id, ConnectionError> {
+        if self.id_counter >= u32::MAX as usize - 2 {
+            return Err(ConnectionError::NoMoreStreamIds)
+        }
+        let proposed = stream::Id::new(self.id_counter as u32);
+        self.id_counter += 2;
+        match self.mode {
+            Mode::Client => assert!(proposed.is_client()),
+            Mode::Server => assert!(proposed.is_server())
+        }
+        Ok(proposed)
+    }
+
+    fn is_valid_remote_id(&self, id: stream::Id, ty: Type) -> bool {
+        match ty {
+            Type::Ping | Type::GoAway => return id.is_session(),
+            _ => {}
+        }
+        match self.mode {
+            Mode::Client => id.is_server(),
+            Mode::Server => id.is_client()
+        }
+    }
+
+    fn new_stream(&mut self, id: stream::Id, recv_window: u32) -> Stream {
+        let recv_win = Arc::new(Window::new(AtomicUsize::new(recv_window as usize)));
+        let (stream_tx, stream_rx) = mpsc::unbounded();
+        let inbox = StreamHandle {
+            recv_win: recv_win.clone(),
+            sender: stream_tx,
+            ack: true
+        };
+        self.streams.insert(id, inbox);
+        Stream::new(id, self.config.clone(), self.stream_tx.clone(), stream_rx.fuse(), recv_win)
+    }
+
+    fn deliver(&mut self, id: stream::Id, item: Item) -> bool {
+        if let Some(ref inbox) = self.streams.get(&id) {
+            if inbox.sender.unbounded_send(item).is_ok() {
+                return true
+            }
+        }
+        trace!("can not deliver; stream {} is gone", id);
+        self.streams.remove(&id);
+        false
+    }
+
+
+    fn terminate(&mut self) {
+        debug!("terminating connection");
+        self.is_dead = true;
+        self.streams.clear()
+    }
+
+    fn send(&mut self, frame: RawFrame) -> Poll<(), ConnectionError> {
+        trace!("send: {:?}", frame);
+        match self.resource.start_send(frame) {
+            Ok(AsyncSink::Ready) => Ok(Async::Ready(())),
+            Ok(AsyncSink::NotReady(frame)) => {
+                self.pending = Some(frame);
```

Comment:
I think this should have a `debug_assert` that `self.pending` is not `None`, so that we can possible bugs.